src/tools/kotlin-schema-generator.ts

```diff
@@ -8,83 +8,102 @@
 
   const schemaNames = schema.names.map(n => `SchemaName("${n}")`);
   const {refinement, query} = generatePredicates(schema);
 
-  function generateFieldMap(isCollection: boolean): string {
-    const fieldPairs = descriptor.fields
-        .filter(f => !!f.isCollection === isCollection)
-        .map(({field, schemaType}) => `"${field}" to ${schemaType}`);
-    return leftPad(ktUtils.mapOf(fieldPairs, 30), 8, true);
-  }
+  const singletons: string[] = [];
+  const collections: string[] = [];
+
+  await visitSchemaFields(schema, ({isCollection, field, schemaType}) =>
+      (isCollection ? collections : singletons).push(`"${field}" to ${schemaType}`));
 
   return `\
 Schema(
     setOf(${ktUtils.joinWithIndents(schemaNames, {startIndent: 8})}),
     SchemaFields(
-        singletons = ${generateFieldMap(false)},
-        collections = ${generateFieldMap(true)}
+        singletons = ${leftPad(ktUtils.mapOf(singletons, 30), 8, true)},
+        collections = ${leftPad(ktUtils.mapOf(collections, 30), 8, true)}
     ),
-    ${quote(descriptor.node.hash)},
+    ${quote(await schema.hash())},
     refinement = ${refinement},
     query = ${query}
 )`;
 }
 
+interface SchemaField {
+  field: string;
+  isCollection: boolean;
+  schemaType: string;
+}
+
+async function visitSchemaFields(schema: Schema, visitor: (field: SchemaField) => void) {
+  for (const [field, descriptor] of Object.entries(schema.fields)) {
+    switch (descriptor.kind) {
+      case 'schema-collection':
+        visitor({field, isCollection: true, schemaType: await getSchemaType(field, descriptor.schema)});
+        break;
+      case 'schema-primitive':
+      case 'kotlin-primitive':
+      case 'schema-reference':
+      case 'schema-nested':
+        visitor({field, isCollection: false, schemaType: await getSchemaType(field, descriptor)});
+        break;
+      case 'schema-ordered-list':
+        visitor({field, isCollection: false, schemaType: await getSchemaType(field, {
+          ...descriptor,
+          innerType: await getSchemaType(field, descriptor.schema)
+        })});
+        break;
+      default:
+        throw new Error(`Schema kind '${descriptor.kind}' for field '${field}' is not supported`);
+    }
+  }
+}
+
+async function getSchemaType(field: string, {kind, schema, type, innerType}): Promise<string> {
+  if (kind === 'schema-primitive') {
+    switch (type) {
+      case 'Text': return 'FieldType.Text';
+      case 'URL': return 'FieldType.Text';
+      case 'Number': return 'FieldType.Number';
+      case 'Boolean': return 'FieldType.Boolean';
+      default: break;
+    }
+  } else if (kind === 'kotlin-primitive') {
+    switch (type) {
+      case 'Byte': return 'FieldType.Byte';
+      case 'Short': return 'FieldType.Short';
+      case 'Int': return 'FieldType.Int';
+      case 'Long': return 'FieldType.Long';
+      case 'Char': return 'FieldType.Char';
+      case 'Float': return 'FieldType.Float';
+      case 'Double': return 'FieldType.Double';
```

Comment:
Looks like `BigInt` (on the Kotlin side) is missing.

Reply:
This PR is only restructuring things so that we can fix the problems with store getting wrong type and to unblock Ray, we are not adding new features here. You can reach out to Shane to ask him about big int.